planning/obstacle_avoidance_planner/src/node.cpp

```diff
@@ -1295,17 +1299,37 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
     const bool is_outside = drivable_area_utils::isOutsideDrivableAreaFromRectangleFootprint(
       traj_point, left_bound, right_bound, vehicle_param_, is_considering_footprint_edges_);
 
-    // only insert zero velocity to the first point outside drivable area
+    // only insert zero velocity to the first point outside drivable area (minus stop margin)
     if (is_outside) {
-      traj_points[i].longitudinal_velocity_mps = 0.0;
+      auto position_outside_drv_area_arc_length =
+        motion_utils::calcSignedArcLength(traj_points, 0, i);
+      size_t stop_idx = i;
+      for (size_t j = i - 1; i > 0; --j) {
+        auto stop_position_arc_length = motion_utils::calcSignedArcLength(traj_points, 0, j);
+        if (
+          stop_position_arc_length <
+          (position_outside_drv_area_arc_length - vehicle_stop_margin_)) {
+          stop_idx = j;
+          break;
+        }
+      }
```

Comment:
Also of course you could use:

```suggestion
const auto & stop_idx = i - std::floor(vehicle_stop_margin_/ delta_arc_length_for_mpt_points) - 1
```

But given that distance between the traj points are not precisely `delta_arc_length_for_mpt_points` we can go with the idea as above.